specs/phase1/beacon-chain.md

```diff
@@ -752,298 +421,108 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
-        get_current_epoch(state),
+    pending_header.votes = bitwise_or(
+        pending_header.votes,
+        attestation.aggregation_bits
     )
 
-    return Validator(
-        pubkey=deposit.data.pubkey,
-        withdrawal_credentials=deposit.data.withdrawal_credentials,
-        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
-        activation_epoch=FAR_FUTURE_EPOCH,
-        exit_epoch=FAR_FUTURE_EPOCH,
-        withdrawable_epoch=FAR_FUTURE_EPOCH,
-        effective_balance=effective_balance,
-        next_custody_secret_to_reveal=next_custody_secret_to_reveal,
-        all_custody_secrets_revealed_epoch=FAR_FUTURE_EPOCH,
+    # Check if the PendingShardHeader is eligible for expedited confirmation
+    # Requirement 1: nothing else confirmed
+    all_candidates = [
+        c for c in pending_headers if
+        (c.slot, c.shard) == (pending_header.slot, pending_header.shard)
+    ]
+    if True not in [c.confirmed for c in all_candidates]:
+        # Requirement 2: >= 2/3 of balance attesting
+        participants = get_attesting_indices(state, attestationg.data, pending_commitment.votes)
+        participants_balance = get_total_balance(state, participants)
+        full_committee = get_beacon_committee(state, attestationg.data.slot, attestationg.data.shard)
+        full_committee_balance = get_total_balance(state, full_committee)
+        if participants_balance * 3 > full_committee_balance * 2:
+            pending_header.confirmed = True
+```
+
+#### `process_shard_header`
+
+```python
+def process_shard_header(state: BeaconState,
+                         signed_header: Signed[ShardDataHeader]) -> None:
+    header = signed_header.message
+    header_root = hash_tree_root(header)
+    # Verify signature
+    signer_index = get_shard_proposer_index(state, header.slot, header.shard)
+    assert bls.Verify(
+        state.validators[signer_index].pubkey,
+        compute_signing_root(header, get_domain(state, DOMAIN_SHARD_HEADER)),
+        signed_header.signature
+    )
+    # Verify length of the header, and simultaneously verify degree.
+    assert (
+        bls.Pairing(header.length_proof, SIZE_CHECK_POINTS[header.commitment.length]) ==
+        bls.Pairing(header.commitment.point, G2_SETUP[-header.commitment.length]))
     )
+    # Get the correct pending header list
+    if compute_epoch_at_slot(header.slot) == get_current_epoch(state):
+        pending_headers = state.current_epoch_pending_shard_headers
+    else:
+        pending_headers = state.previous_epoch_pending_shard_headers
+        
+    # Check that this header is not yet in the pending list
+    for pending_header in pending_headers:
+        assert header_root != pending_header.root
+    # Include it in the pending list
+    committee_length = len(get_beacon_committee(state, header.slot, header.shard))
+    pending_headers.append(PendingShardHeader(
+        slot=header.slot,
+        shard=header.shard,
+        commitment=header.commitment,
```

Comment:
```suggestion
        commitment=header.commitment.point,
        length=header.commitment.length,
```

I think we forgot to add the length

Reply:
We actually don't need the length as part of the object, if the `.commitment` property is a `DataCommitment`